### PR TITLE
Add caml_alloc_custom_mem

### PR DESCRIPTION
**Pull request by damiendoligez**
This PR is related to #1476 and [MPR#7750](https://caml.inria.fr/mantis/view.php?id=7750).

It adds one function to the runtime and three new GC parameters.

The function (`caml_alloc_custom_mem`) is to be used by C code to allocate custom objects that consist of a small block in the heap pointing to a large amount of memory outside the heap (typically allocated with `malloc`). This function takes only one size parameter, the number of bytes of memory outside the heap used by this custom block.

The user can then control how much of this memory is retained by dead blocks because the GC does not immediately collect dead blocks ("floating garbage"). This is done with three new GC parameters:

* `M` `custom_major_ratio` This is a percentage relative to the major heap size. A complete GC cycle will be done every time 2/3 of that much memory is allocated for blocks in the major heap. Assuming constant allocation and deallocation rates, this means there are at most (M/100 * major-heap-size) bytes of floating garbage at any time. The factor of 2/3 (or 1.5) is (roughly speaking) because the major GC takes 1.5 cycles (previous cycle + marking phase) before it starts to deallocate dead blocks allocated during the previous cycle.

* `m` `custom_minor_ratio` This is a percentage relative to the minor heap size. A minor GC will be forced when the memory allocated for blocks in the minor heap reaches this value.

* `n` `custom_minor_max_size` This is a number of bytes. Blocks allocated in the minor heap that are over this size are assumed to be long-lived and counted immediately against the major heap rather than the minor heap. This is to avoid large allocations forcing a minor GC every time.

The function is then used to allocate bigarrays, allowing us to get rid of the infamous `CAML_BA_MAX_MEMORY` constant.

When applied to the test program of [MPR#7750](https://caml.inria.fr/mantis/view.php?id=7750), the results are pretty good: we can make the process size vary from about 6GB (with M=10) to 15 GB (with M=100) to 25 GB (with M=200).

**needs to be done**: documentation

### CODE REVIEW

**Comment by alainfrisch** [comment]
This should also hopefully address MPR#7100.

File: byterun/major_gc.c [hunk 1]
```diff
@@ -693,7 +694,12 @@ void caml_major_collection_slice (intnat howmuch)
   }
   if (p < dp) p = dp;
   if (p < caml_extra_heap_resources) p = caml_extra_heap_resources;
-  if (p > 0.3) p = 0.3;
+  p += p_backlog;
+  p_backlog = 0.0;
```
Comment (alainfrisch) [inline byterun/major_gc.c]: Can you explain this change?  Is it strictly related to the new API, or could it impact the behavior of programs not using it?

**Comment by alainfrisch** [comment]
Not strictly speaking related to the proposed change, but I think it's good to maintain an overall understanding of how custom blocks are handled by the GC.  Could you explain the following heuristics in caml_adjust_gc_speed:

````c
  if (caml_extra_heap_resources
           > (double) caml_minor_heap_wsz / 2.0
             / (double) caml_stat_heap_wsz) {
    CAML_INSTR_INT ("request_major/adjust_gc_speed_2@", 1);
    caml_request_major_slice ();
  }
````

As the heap grows during the life of the program, this will trigger more and more often.  For instance, each i/o channel allocation (currently with ratio 1/1000) will trigger a major slice as soon as:
````
  caml_stat_heap_wsz > 500 * caml_minor_heap_wsz
````

and this can easily happen.  Does it really make sense to trigger GC slices more often when the heap grows for a constant allocation rate of custom blocks?

File: byterun/major_gc.c (hunk 1 above)
Comment (damiendoligez) [reply byterun/major_gc.c]: This is suggested in the last line of [MPR#7180](https://caml.inria.fr/mantis/view.php?id=7180) and repeated in [MPR#7750](https://caml.inria.fr/mantis/view.php?id=7750). If the extra_heap_resources ask for a huge slice (larger than one complete phase), we should probably not do just one phase, then forget about the rest.

I must confess I haven't tested the effects of this change in isolation.

To answer your second question, it does impact the behavior of programs that don't use the new API, hopefully in a positive way.

**Comment by damiendoligez** [comment]
A bit of digging with `git blame` reveals that this heuristic was already present in 1995 in the very first commit in OCaml's history, so it dates back to Caml Light... It's probably time to re-assess because I really don't remember the reason for this.

**Comment by xavierleroy** [comment]
On Thu, Apr 26, 2018 at 4:10 PM Damien Doligez <notifications@github.com>
wrote:

> A bit of digging with git blame reveals that this heuristic was already
> present in 1995 in the very first commit in OCaml's history, so it dates
> back to Caml Light... It's probably time to re-assess because I really
> don't remember the reason for this.
>
If you're in the mood for archeology, the Caml Light sources (with CVS
history) are here: https://github.com/camllight/camllight

Enjoy :-)

- Xavier

**Comment by stedolan** [comment]
This looks really nice. The only bit that worries me is `custom_minor_max_size`, because it looks like performance might change abruptly when allocation sizes go from 8191 to 8193 bytes. Rather than a threshold, how about something smoother, like say adding the first 8k to `extra_heap_resources_minor` and the rest to `extra_heap_resources`?

**Comment by alainfrisch** [comment]
> If you're in the mood for archeology, the Caml Light sources (with CVS
history) are here

This leads to :

https://github.com/camllight/camllight/commit/67c1dc3d526dda8a9cb0d33116538212f5e45101#diff-d0f533866c552cc04626e6d892579278R153

which does not give any rationale for that specific heuristics.

**Comment by damiendoligez** [comment]
> This looks really nice. The only bit that worries me is custom_minor_max_size, because it looks like performance might change abruptly when allocation sizes go from 8191 to 8193 bytes. Rather than a threshold, how about something smoother, like say adding the first 8k to extra_heap_resources_minor and the rest to extra_heap_resources?

Sounds like a good idea. I'll try it.

**Comment by damiendoligez** [comment]
@stedolan Done.

**Comment by alainfrisch** [comment]
@damiendoligez Should this new API also be used for i/o channels?

File: byterun/major_gc.c (hunk 1 above)
Comment (alainfrisch) [reply byterun/major_gc.c]: Ok it makes sense, thanks for the explanation.

**Comment by xavierleroy** [comment]
@alainfrisch: I/O buffers consume both one file descriptor (a system resource, perhaps more appropriate for the old API?) and an out-of-heap buffer (a memory resource, very appropriate for the new API).  So, it's a tie.

This said I'm very happy to see the new API.  If the experts are OK with it I suggest merging in trunk ASAP.

**Comment by alainfrisch** [comment]
>  I/O buffers consume both one file descriptor (a system resource, perhaps more appropriate for the old API?)

It don't think it's the case.  The life-time of the file descriptor does not depend on the GC: it can and must be released explicitly (with close), and the GC finaliser won't touch it (no implicit close).  So as far as I can tell, the channel object is equivalent to a pure memory buffer, as far as the GC is concerned.

**Comment by alainfrisch** [comment]
(About i/o channels:  with an extra indirection, one could even destroy (free) the buffer when the channel is (explicitly) closed, but it's tricky to make this thread-safe.)

**Comment by damiendoligez** [comment]
@alainfrisch you're right, the FD part of the channel is unmanaged, and the memory part is a good fit for the new API, so I'm going to convert it too.

Don't merge before I do that.

**Comment by delamonpansie** [comment]
Why just not use `caml_alloc_dependent_memory` and `caml_free_dependent_memory`?
Yes, this would require extending caml_ba_proxy with field holding actual size of allocation. It simple and relies on already existing Gc infrastructure.

Here is a small test case: (it requires 4.07)
```
module Gc_helper = struct
  type void
  external alloc_dependent_memory : (int [@untaged]) -> void = "caml_alloc_dependent_memory"
  external free_dependent_memory : (int [@untaged]) -> void = "caml_free_dependent_memory"

  let allocated = ref 0
  let minor_heap_size = Gc.((get ()).minor_heap_size) / (Sys.word_size / 8)

  let bigarray v =
    let size = Bigarray.(Array1.size_in_bytes v) in
    allocated := !allocated + size;
    if !allocated > minor_heap_size then begin
        allocated := size;
        Gc.major_slice (-1) |> ignore; (* heuristics from https://github.com/ocaml/ocaml/blob/trunk/byterun/memory.c#L522 *)
      end;
    alloc_dependent_memory size |> ignore;
    Gc.finalise_last (fun () -> free_dependent_memory size |> ignore) v; (* this should go to caml_ba_finaliser *)
    v
end

module Memstat = struct
  let word_size = Sys.word_size / 8 (* word size in bytes *)

  let kmgt size =
    let sprintf = Printf.sprintf in
    if size < 512 * 1024 then
      sprintf "%.2fk" (float size /. 0x1.0p10)
    else if size < 512 * 1024 * 1024 then
      sprintf "%.2fm" (float size /. 0x1.0p20)
    else if size < 512 * 1024 * 1024 * 1024 then
      sprintf "%.2fg" (float size /. 0x1.0p30)
    else
      sprintf "%.2ft" (float size /. 0x1.0p40)

  let statm () =
    let open Scanf in
    let fd = Scanning.from_file "/proc/self/statm" in
    let result = bscanf fd "%i %i %i"
                   (fun size resident shared -> (size, resident, shared) ) in
    Scanning.close_in fd;
    result

  let alive_buf_size = ref 0
  let total_buf_size = ref 0

  let alloc size v =
    total_buf_size := !total_buf_size + size;
    alive_buf_size := !alive_buf_size + size;
    Gc.finalise_last (fun () -> total_buf_size := !total_buf_size - size) v
  let free size =
    alive_buf_size := !alive_buf_size - size

  let rec memstat () =
    let stat = Gc.stat () in
    let vss, _, _ = statm () in
    Printf.printf "% 9s vss; % 9s heap; % 9s live; % 9s alive buf; % 9s total buf; % 9.0f%% garbage\n%!"
      (kmgt (vss * 4096))
      (kmgt (stat.heap_words * word_size))
      (kmgt (stat.live_words * word_size))
      (kmgt !alive_buf_size)
      (kmgt !total_buf_size)
      (100. *. float (!total_buf_size - !alive_buf_size) /. float !alive_buf_size);
    Unix.sleepf 0.5;
    memstat ()

  let _ =
    Thread.create memstat ()
end


(* Create some memory pressure, simulating long running programm with some
   reasonable amount of small objects *)
let tree =
  let limit = 1024 * 1024 * (try int_of_string (Sys.getenv "LIMIT")
                             with _ -> 10) in
  let key size =
    Bytes.init size (fun _ -> Random.int 9 + Char.code '0' |> Char.chr) in
  let module T = Set.Make(Bytes) in
  let rec build total_size t =
    if total_size > limit then
      t
    else
      let size = Random.int 16 in
      build (total_size + size) (T.add (key size) t)
  in
  print_string "Allocating some objects\n";
  build 0 T.empty

let run message alloc =
  Gc.compact ();
  print_string message;
  print_newline ();

  let q = Queue.create () in
  let deadline = Unix.gettimeofday () +. 15.0 in
  while Unix.gettimeofday () < deadline do
    let size = 2 * 1024 * 1024 + Random.int (2 * 1024 * 1024) in
    let buf = alloc size in
    Queue.push (size, buf) q;
    Memstat.alloc size buf;

    (* pretend  we have about 30 active buffers at every moment of time *)
    while Queue.length q > 30 do
      let size, _ = Queue.pop q in
      Memstat.free size;
    done;
  done;

  (* drain queue *)
  while Queue.length q > 0 do
    let size, _ = Queue.pop q in
    Memstat.alive_buf_size := !Memstat.alive_buf_size - size
  done


let _ =
  run "Bytes" (fun size -> Bytes.create size);
  run "helped Bigarray" (fun size -> Bigarray.(Array1.create char c_layout size) |> Gc_helper.bigarray);
  run "plain Bigarray" (fun size -> Bigarray.(Array1.create char c_layout size));


(* Local Variables: *)
(* compile-command: "ocamlopt -thread unix.cmxa threads.cmxa bigarray.cmxa gc_sim.ml -o gc_sim" *)
(* End: *)

**Comment by damiendoligez** [comment]
> Why just not use caml_alloc_dependent_memory and caml_free_dependent_memory?

That API is harder to use (it mandates the use of finalizers) and never caught on. It doesn't let the user control the ratio of heap to external memory.

**Comment by gasche** [comment]
As far as I can tell, this PR has not been fully reviewed (and not approved), and @damiendoligez indicates that it is ready. @stedolan, would you maybe be willing to do a review?

**Comment by gasche** [comment]
( @alainfrisch or @jhjourdan might be qualified to review as well. )

File: man/ocamlrun.m [hunk 2]
```diff
@@ -125,6 +125,7 @@ in the OCAMLRUNPARAM environment variable (see below).
 record documented in
 .IR "The OCaml user's manual",
 chapter "Standard Library", section "Gc".
+\" FIXME missing: c, H, w, W
```
Comment (alainfrisch) [inline man/ocamlrun.m]: (This is unrelated to PR.  Perhaps open a Mantis issue?)

File: man/ocamlrun.m [hunk 3]
```diff
@@ -168,6 +169,38 @@ The default size increment for the major heap (in words).
...
+held by custom values in the minor heap. A minor GC is triggered
+when this much memory is held by custom values located in the minor
+heap. Expressed as a percentage of minor heap size. Default:
+100. Note: this only applies to values allocated with
+.B caml_alloc_custom_mem
+(e.g. bigarrays).
+.TP
+.BR n \ (custom_minor_max_size)
+Maximum amount of out-of-heap
+memory for custom values allocated in the minor heap. When a custom
```
Comment (alainfrisch) [inline man/ocamlrun.m]: I'd say "for each custom value allocated in the minor heap".

File: man/ocamlrun.m [hunk 4]
```diff
@@ -168,6 +169,38 @@ The default size increment for the major heap (in words).
 .BR l \ (stack_limit)
 The limit (in words) of the stack size.
 .TP
+.BR M \ (custom_major_ratio)
+Target ratio of floating garbage to
+major heap size for out-of-heap memory held by custom values
+located in the major heap. The GC speed is adjusted
+to try to use this much memory for dead values that are not yet
+collected. Expressed as a percentage of major heap size. Default:
+30. Note: this only applies to values allocated with
```
Comment (alainfrisch) [inline man/ocamlrun.m]: 44, no?

File: man/ocamlrun.m [hunk 5]
```diff
@@ -168,6 +169,38 @@ The default size increment for the major heap (in words).
...
+.TP
+.BR n \ (custom_minor_max_size)
+Maximum amount of out-of-heap
+memory for custom values allocated in the minor heap. When a custom
+value is allocated   on the minor heap and holds more than this many
+bytes, only this value is counted against
+.B custom_minor_ratio
+and the rest is directly counted against
+.BR custom_major_ratio .
+Note:
```
Comment (alainfrisch) [inline man/ocamlrun.m]: Mention the default (8kB)?

File: manual/manual/cmds/runtime.etex [hunk 6]
```diff
@@ -157,6 +157,27 @@ The following environment variables are also consulted:
...
+  when this much memory is held by custom values located in the minor
+  heap. Expressed as a percentage of minor heap size. Default:
+  100. Note: this only applies to values allocated with
+  "caml_alloc_custom_mem".
+  \item[n] ("custom_minor_max_size") Maximum amount of out-of-heap
+  memory for custom values allocated in the minor heap. When a custom
+  value is allocated on the minor heap and holds more than this many
+  bytes, only this value is counted against "custom_minor_ratio" and
+  the rest is directly counted against "custom_major_ratio". Note:
+  this only applies to values allocated with "caml_alloc_custom_mem".
```
Comment (alainfrisch) [inline manual/manual/cmds/runtime.etex:181]: Same comments as for the man page.

File: runtime/custom.c [hunk 7]
```diff
@@ -20,32 +20,47 @@
...
+  CAMLassert (mem_minor <= mem);
+
+  wosize = 1 + (bsz + sizeof(value) - 1) / sizeof(value);
   if (wosize <= Max_young_wosize) {
     result = caml_alloc_small(wosize, Custom_tag);
     Custom_ops_val(result) = ops;
     if (ops->finalize != NULL || mem != 0) {
-      /* Remember that the block needs processing after minor GC. */
-      add_to_custom_table (&caml_custom_table, result, mem, max);
+      if (mem > mem_minor){
```
Comment (alainfrisch) [inline runtime/custom.c]: nitpick:
```suggestion
      if (mem > mem_minor) {
```

File: stdlib/gc.mli [hunk 8]
```diff
@@ -149,7 +149,30 @@ type control =
...
+
+    custom_major_ratio : int;
+    (** A percentage relative to the major heap size. This parameter
+        controls the major GC speed as a function of memory allocated
+        outside the heap and retained by custom values located in the
+        major heap. The GC will speed up as needed to (try to) limit
+        the amount of floating garbage to this much memory. The
+        default value keeps the out-of-heap floating garbage about the
+        same size as the in-heap overhead.
+        Default: 44. @since 4.08.0 *)
```
Comment (alainfrisch) [inline stdlib/gc.mli]: Don't you want to mention here that these new settings only apply to custom values allocated using the new API?

File: testsuite/tests/misc/ephetest2.ml [hunk 9]
```diff
@@ -144,9 +144,10 @@ let run test init =
   Stack.clear env.varephe_false;
   Gc.full_major ();
   let res = Hashtbl.fold (fun _ v acc -> acc && check_var v) env.vars true in
-  is_true test "check" res
+  is_true test "check" res;
+  env (* Keep env.varephe_true alive. *)
```
Comment (alainfrisch) [inline testsuite/tests/misc/ephetest2.ml:148]: Is this change made necessary by this PR, or did it just happen by chance that the problem was not triggered before?

File: stdlib/gc.mli [hunk 10]
```diff
@@ -149,7 +149,30 @@ type control =
         out variations in its workload. This is an integer between
         1 and 50.
         Default: 1. @since 4.03.0 *)
-}
+
+    custom_major_ratio : int;
+    (** A percentage relative to the major heap size. This parameter
+        controls the major GC speed as a function of memory allocated
```
Comment (alainfrisch) [inline stdlib/gc.mli]: Why not reuse the text from the manual ("Target ratio of floating garbage to
major heap size for out-of-heap memory held by custom values
located in the major heap...")?

**Review by alainfrisch — APPROVED** [review]
LGTM, only cosmetic comments.

@damiendoligez Did you test [MPR#7100](https://caml.inria.fr/mantis/view.php?id=7100) against this branch?

**Comment by alainfrisch** [comment]
You could also use the new API in win32graph/draw.c, graph/image.c and avoid magic constants in those files.  It seems all other uses in the core distribution are with default values for mem/max (0/1), so could also be switched to new API (or not).

File: testsuite/tests/misc/ephetest2.ml (hunk 9 above)
Comment (damiendoligez) [reply testsuite/tests/misc/ephetest2.ml:148]: I would say that was a latent bug in the test program.

File: man/ocamlrun.m (hunk 2 above)
Comment (damiendoligez) [reply man/ocamlrun.m]: Done in [MPR#7870](https://caml.inria.fr/mantis/view.php?id=7870). Any objection to also keeping the FIXME here?

File: man/ocamlrun.m (hunk 3 above)
Comment (damiendoligez) [reply man/ocamlrun.m]: Right.

File: man/ocamlrun.m (hunk 4 above)
Comment (damiendoligez) [reply man/ocamlrun.m]: Right.

File: man/ocamlrun.m (hunk 5 above)
Comment (damiendoligez) [reply man/ocamlrun.m]: Right.

File: stdlib/gc.mli (hunk 8 above)
Comment (damiendoligez) [reply stdlib/gc.mli]: Right.

File: stdlib/gc.mli (hunk 10 above)
Comment (damiendoligez) [reply stdlib/gc.mli]: Done.

File: man/ocamlrun.m (hunk 2 above)
Comment (alainfrisch) [reply man/ocamlrun.m]: No, of course.

**Comment by alainfrisch** [comment]
@damiendoligez : careful, I think something went wrong with a merge/rebase on the branch.

**Comment by damiendoligez** [comment]
I fixed the rebase problem. Now I'll look at graph and other uses in the core distribution.

**Comment by damiendoligez** [comment]
I'm not sure about Win32, but on X Window, the memory held by a Pixmap is in the server, not in the user process, so I don't think it makes sense to use the new API there.

For the other uses, you're right that they all use 0/1 so converting them wouldn't change anything.

**Comment by damiendoligez** [comment]
I tried the test program of [MPR#7100](https://caml.inria.fr/mantis/view.php?id=7100) and indeed with this PR it doesn't fail.

File: stdlib/gc.mli [hunk 11]
```diff
@@ -149,7 +149,42 @@ type control =
         out variations in its workload. This is an integer between
         1 and 50.
         Default: 1. @since 4.03.0 *)
-}
+
+    custom_major_ratio : int;
+
```
Comment (alainfrisch) [inline stdlib/gc.mli]: This empty line explains the CI failure:
````
File "/home/travis/build/ocaml/ocaml/stdlib/gc.mli", line 155, characters 4-580:
Error (warning 50): unattached documentation comment (ignored)
````

Comment (damiendoligez) [reply stdlib/gc.mli]: Yes yes, I'm fixing it.

**Comment by damiendoligez** [comment]
Note: it will be time to revisit #1961 after this PR is merged and tested in the wild for some time.

**Comment by alainfrisch** [comment]
> I tried the test program of MPR#7100 and indeed with this PR it doesn't fail.

(Actually, I'm confused on whether #1476 already fixed that one.)

**Review by alainfrisch — APPROVED** [review]
LGTM

**Comment by damiendoligez** [comment]
> (Actually, I'm confused on whether #1476 already fixed that one.)

I don't think it did. I tried the test program with 4.07.1 and it failed.

**Comment by damiendoligez** [comment]
CI passed, so I'm merging. Thanks for the review!

**Comment by alainfrisch** [comment]
Other tickets more or less directly related to this PR:

https://caml.inria.fr/mantis/view.php?id=7198
https://caml.inria.fr/mantis/view.php?id=7676 (somehow)
https://caml.inria.fr/mantis/view.php?id=6294
https://caml.inria.fr/mantis/view.php?id=7158

(I'm not sure it's worth mentioning them in the Changelog, but I'll let @damiendoligez decide.)

**Comment by damiendoligez** [comment]
This PR resolves [MPR#7198](https://caml.inria.fr/mantis/view.php?id=7198) but the others will also need changes in other software.

I'll add it to Changes.

**Comment by damiendoligez** [comment]
I added a mention of MPR#7198 (3f46a1d9233ceecdd55898e316cfc2d963855ce2). The others are not directly fixed by this PR.
